### PR TITLE
Follow spec for all event authentication

### PR DESCRIPTION
**Pull request by DevinR528**
So far this makes `is_membership_change_allowed` follow the spec plus a federation check.

### CODE REVIEW

File: src/event_auth.rs [hunk 1]
```diff
@@ -186,13 +190,17 @@ pub fn auth_check(
...
+        //     tracing::warn!("state_key must be non-empty");
+        //     return Ok(false); // and be non-empty state_key (point to a user_id)
+        // }
+
+        // TODO what? "sender's domain doesn't matches"
+        // If sender's domain doesn't matches state_key, reject
         if event.state_key() != Some(event.sender().to_string()) {
-            tracing::warn!("no state_key field found for event");
+            tracing::warn!("state_key does not match sender");
             return Ok(false);
```
Comment (DevinR528) [inline src/event_auth.rs]: @timokoesters, is this what they mean for m.room.aliases or do they mean `sender().server_name()`?

https://matrix.org/docs/spec/rooms/v1#authorization-rules

File: src/event_auth.rs [hunk 2]
```diff
@@ -637,27 +675,31 @@ fn get_deserialize_levels(
...
         {
+            tracing::info!("redaction event allowed via room version 1 rules");
             return Ok(RedactAllowed::OwnEvent);
         }
-    } else {
-        // TODO synapse has this line also
-        // event.internal_metadata.recheck_redaction = True
+    // redactions to events where the sender's domains match, allow
+    } else if redacted_event.map(|ev| ev.sender()) == Some(redaction_event.sender()) {
+        tracing::info!("redaction allowed via own redaction");
```
Comment (DevinR528) [inline src/event_auth.rs:694]: Same here for m.room.redaction am I reading this right?

https://matrix.org/docs/spec/rooms/v3#authorization-rules-for-events

File: src/event_auth.rs (hunk 1 above)
Comment (timokoesters) [reply src/event_auth.rs]: You should compare event.state_key() with event.sender().server_name().as_str()

File: src/event_auth.rs (hunk 2 above)
Comment (timokoesters) [reply src/event_auth.rs:694]: > redaction events are always accepted (provided the event is allowed by events and events_default in the power levels)

Comment (timokoesters) [reply src/event_auth.rs:694]: You should remove this check

**Pull request merged**